### DISCUSSION TITLE
SwiperContainer and SwiperSlide types in swiper-element

### DISCUSSION THREAD

**Discussion by konclave** [nolimits4web/swiper#6394] [Q&A]
I'm using Swiper 9.0.4 element components in React app and trying to set the typings. 
I've added custom elements to `JSX IntrinsicElements` interface, but I'm having issues with typing those custom elements:
```typescript
import { SwiperContainer, SwiperSlide } from 'swiper/element';

declare global {
  namespace JSX {
    interface IntrinsicElements {
      'swiper-container': React.DetailedHTMLProps<React.HTMLAttributes<typeof SwiperContainer>, typeof SwiperContainer>;
      'swiper-slide': React.DetailedHTMLProps<React.HTMLAttributes<typeof SwiperSlide>, typeof SwiperSlide>;
    }
  }
}
```

As I see in the Swiper library, it provides these types: 
```typescript
declare const SwiperContainer: HTMLElement;
declare const SwiperSlide: HTMLElement;
```

but if I set it as a type for custom elements, then it lacks the properties and methods that were implemented in the custom element: 
```jsx
<swiper-container ref={swiperRef} init={false}> // here I have a typescript error: Property 'init' does not exist on type 'DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>'.
...
</swiper-container>
```
and when I try to call the methods of custom element, I also have an error that the method doesn't exist on type HTMLElement:

```typescript
swiperRef.current.initialize();
```

How do I handle this typing errors? Why `SwiperContainer` and `SwiperSlide` types are exposed as HTMLElement. 
Seems like a bug for me.

**Comment by tehmann83** [comment]
Importing `SwiperCore` fixed this for me (using **version 8.4.7**)

```
import SwiperCore from 'swiper';

// ...


const [swiperInstance, setSwiperInstance] = useState<SwiperCore>();

// ... set your swiper to the swiperInstance state

```

This did the trick to get back properties, methods, and getting rid of the type errors.

**Reply by konclave** [reply]
It does not work in my case, because I need to add types to `SwiperContainer` and `SwiperSlide` custom elements, and they have an interface which is different from `SwiperCore`.

**Reply by rohanbhangui** [reply]
having a similar issue when trying to use the swiper-elements as well

**Reply by tehmann83** [reply]
Have you tried extending the `SwiperCore` types with the custom types?

```
interface CustomStuff extends SwiperCore{
   customElement: string; 
}
```

**Reply by konclave** [reply]
I can create the custom element type by myself, but I expect that it to be exported from the library.
